encoder.go

```diff
@@ -25,12 +25,33 @@ import (
 	"time"
 )
 
-// Encoder is a format-agnostic interface for all log field encoders. It's not
-// safe for concurrent use.
+// Encoder is a format-agnostic interface for all log entry marshalers. Since
+// log encoders don't need to support the same wide range of use cases as
+// general-purpose marshalers, it's possible to make them much faster and
+// lower-allocation.
+//
+// Encoders need not be fully safe for concurrent use, but it should be safe to
```

Comment:
Now that this is documented, it seems very strange to require some methods to work concurrently, but not others. Should we actually have 2 interfaces here that separate a mutable state compared to the immutable state that is safe for concurrent use? E.g., breaking out the `WriteEntry` into a separate interface that is safe for concurrent use, while adding fields and `Clone` are not safe for concurrent use.